### DISCUSSION TITLE
is there a tool to control strength of S_NOISE or ETA via mask

### DISCUSSION THREAD

**Discussion by AlexD81** [Comfy-Org/ComfyUI#4520] [Q&A] [answered]
Hi, is there a tool to control strength of S_NOISE or ETA via mask/map? Need to reduce/increase S_NOISE for some parts of image. Of course, it can be done using mix of multiple ksamplers, but generation time become too high.

---------
for those who interesting about NOISE and ETA, here's some info about them
https://sandner.art/sigma-and-eta-noise-image-variations-in-latent-space-and-fast-detail-effects/

**Comment by blepping** [comment, marked as answer]
@AlexD81 it's possible for `s_noise` (in a slightly roundabout way) with my `SonarCompositeNoise` node. the example actually demonstrates using stronger noise in some areas based on a mask: https://github.com/blepping/ComfyUI-sonar/blob/main/docs/advanced_noise_nodes.md#sonarcompositenoise

i don't think masks with ETA would make sense, ETA doesn't directly control the noise level. it's more like where the step begins and ends (and then based on ETA noise might get added to adjust the level to where a step normally would end). so to do masking with ETA you'd have to step to different sigmas in the actual sampler.

not exactly your question but if you want to have a lot of control over sampling you might also be interested in this project: https://github.com/blepping/comfyui_overly_complicated_sampling (OCS also has integration with ComfyUI-sonar, you can use all the custom noise types from Sonar with it.)

**Reply by AlexD81** [reply]
thank you for such detailed answer!  look pretty interesting,  I'll take a look

**Reply by AlexD81** [reply]
Can I ask you a question pls?   I'm working with landscaping/nature   using img2img with relatively high denoise (0.6) ,    sd1.5
And trees often have regular structure, like radial lines from it's center, it looks computer-generated.  And grass often has "waves  of gaussian noise".  I'm messing with it very long, tried a lot approaches, freeu, ipadapter, samplers, noises,  but still not completely solved it. It's somehing beyound my understanding, maybe you can suggest something,  maybe to use some custom noise, don't know
 
![изображение](https://github.com/user-attachments/assets/56b191ba-b643-4d95-bdf3-6d462b3b4c29)

**Reply by ltdrdata** [reply]
FYI,
https://github.com/Extraltodeus/noise_latent_perlinpinpin
https://github.com/WASasquatch/PowerNoiseSuite
https://github.com/WASasquatch/PPF_Noise_ComfyUI

**Reply by blepping** [reply]
>  And grass often has "waves of gaussian noise".

that's probably not the case. :) you won't be able to see any kind of pattern in Gaussian noise. it's more likely your issue is related to the model and/or LoRAs you're using (possibly other parameters as well). _however_ you might be able to use stuff like weird noise types to force the model out of its normal pattern. it's possible to chain the noise nodes (including the noise compositing nodes) from Sonar so you can mix however many noise types you want and use as many areas as you want (of course they may be a performance cost to generating and mixing many layers/types of noise).

the thing i've learned from my own generations/projects is that trying stuff and experimenting is rewarded. so my advice is to just try some different noise types at various strengths. you can accomplish a partial mix of weird noise + normal noise with something like this:

![image](https://github.com/user-attachments/assets/abb480a7-7fc7-4a34-9e82-98257273ecc5)

`rescale` rebalances the `factor`s of the noise node and the ones connected to it to add up to the rescale value so this is 50% Gaussian noise, 50% Perlin noise. these noise types can be used with any sampler that adds noise during sampling (ancestral and SDE samplers) via the `SamplerConfigOverride` node. you can also use them as the initial noise for generation (or added noise in img2img) with `NoisyLatentLike` (also in Sonar). (`rescale` at 0 just disables rescaling, it doesn't scale everything to 0.)

> FYI, https://github.com/Extraltodeus/noise_latent_perlinpinpin https://github.com/WASasquatch/PowerNoiseSuite https://github.com/WASasquatch/PPF_Noise_ComfyUI

on the subject of Perlin noise, you could possibly also try this branch of OCS: https://github.com/blepping/comfyui_overly_complicated_sampling/tree/perlin3d_experiment

it adds an advanced Perlin node that can generate a cube of 3D Perlin noise. each time noise is needed, it will take a 2D slice through the depth. it has an absurd number of tuneable parameters so you can get many different effects. (if you want to use it with stuff that takes `SONAR_CUSTOM_NOISE` there's also an adapter - not needed for OCS itself).

![image](https://github.com/user-attachments/assets/96196ccf-c30b-4e3e-bcbc-ee0a4aeb034d)

**Reply by AlexD81** [reply]
whank you for such detailed info!    but I'm confused a bit with names.  On this image 
![360571606-abb480a7-7fc7-4a34-9e82-98257273ecc5](https://github.com/user-attachments/assets/4798c129-3131-4859-9ceb-5a4da24a417d)
is rescale  is a kind of "blend"   value?     and 1.0 is blend of 50%.  Checked, it works,  But what sould i set for, let's say,  30% ?    
Tried rescale of 0.7    but result looks like reducing s_noise , so hard to understand how does it works

**Reply by blepping** [reply]
no problem. `factor` is the base strength of the generated noise. generally you want the base strength (which may get multiplied by other stuff later like `s_noise`, the sigma, etc but don't worry about that now) to be `1.0`. one way to deal with when connecting multiple noise nodes is to manually set `factor` such that the `factor`s of the connected nodes adds up to `1.0`.

so in this example, you could set the two `factor`s to `0.5` and you'd get a total of `1.0`. however that can be difficult/time consuming if you have a bunch of noise nodes, which is where the `rescale` parameter comes in. first, if `rescale` is set to `0` that is considered disabled: it won't do any rescaling, you'll just get the factors added together.

when `rescale` is set to a non-zero value then the node with `rescale` enabled _and_ the nodes connected to it will have their `factors` rebalanced to add up to the value in `rescale`. the rebalancing is proportionate so in this example the two `1.0` `factor` parameters would get rebalanced to `0.5`. if you set `factor` in the Perlin noise node to `2.0` and left it as-is in the Gaussian node then the Perlin node's `factor` would get rebalanced to `2/3` and the Gaussian node's `factor` would get rebalanced to `1/3`.

to put it a different less math way, if you connected several nodes like this

```plaintext
factor=1
   |
factor=1
   |
factor=1
   |
factor=2
   |
factor=1 rescale=1
```

you'd know they added up to the correct strength and the `factor=1` nodes would have equal weight relative to each other and the `factor=2` node would have double weight compared to the `factor=1` nodes. the manual way to do it would be to disable `rescale` and set the `factor=1` nodes to `1/6` and the `factor=2` node to `1/3`: `(1 / 6) * 4 + 1 / 3 = 1.0`

the way it works internally is summing the absolute value of the `factor`s and then multiplying each node's `factor` by `rescale / sum_factors`.

> But what sould i set for, let's say, 30% ? Tried rescale of 0.7 but result looks like reducing s_noise , so hard to understand how does it works

no need to use `rescale` for that and you're correct, setting `rescale` to `0.7` would be reducing the strength since the node and the ones it's connected to would have their `factor`s rebalanced to add up to `0.7`. `rescale` is mostly useful when you want to think about the noise type contribution proportionately, like "i want the Gaussian and Pyramid noise types to have equal strength and the Perlin noise type to have proportionately twice as much effect".

hope i didn't just end up confusing you more with this attempted explanation!

**Reply by AlexD81** [reply]
thank you for explanation, but it's appears too complicated especcially for my limited english knowledge. I'll back to this later,  To be honest i'm still messing with my original question.  if i  need to increase/decrease noise in area specified by mask, what parameter should i change? .  The problem is no matter what parameter i change, it also always affect unmasked area,  inverse way.  E.g increasing noise results to decreasing noise in unmasked area.   Tried to enable normalize, but it didnt really help    
![изображение](https://github.com/user-attachments/assets/c1831d3f-1c66-4122-a4c6-038886c91b88)

**Reply by blepping** [reply]
sorry this reply took so long, i must have accidentally marked it as done somehow. 

i just checked at the example workflows still work. i'm going to be referring to the example workflow here: https://github.com/blepping/ComfyUI-sonar/blob/main/docs/advanced_noise_nodes.md#sonarcompositenoise - you can expand the example image in that section and load the "pyramid" one. by the way, all the example images in the ComfyUI-sonar docs are workflow included.

the good news is, you don't have to care about setting rescale here so you can forget all that for right now. in that workflow, the leftmost bottom `SonarCustomNoise` node (the one connected to `sonar_custom_noise_src`) is what applies to the masked area.

Pyramid is a weird noise type so let's use something a bit more normal for this. also, probably want to disable `normalize` in the `SamplerConfigOverride` node (i'm actually not sure why it seems to work without that). i also adjusted the schedule it starts to begin at `0.4` (40% of sampling) instead.

![image](https://github.com/user-attachments/assets/c7301552-b2eb-40d0-9074-2943d202a673)

that just has `factor: 1.0`, so normal strength:

![image](https://github.com/user-attachments/assets/a60b8613-f778-440e-a0e7-83b5235c1a22)

with `factor: 0.5` instead. this is not enough noise so you can see it loses detail in that area:

![image](https://github.com/user-attachments/assets/2053ddc6-c4c5-4e1c-8372-1b28f7727bd4)

with `factor: 1.1` - detail increased in that area:

![image](https://github.com/user-attachments/assets/3a86f178-fafb-4d1f-8ba5-4a8b68707086)

with `factor: 1.2`:

![image](https://github.com/user-attachments/assets/144b2735-4d26-4373-919e-b3d1211807f0)

and finally with `factor: 1.4`:

![image](https://github.com/user-attachments/assets/335d3722-a189-4d66-a618-1737e5bf33c3)

obviously this is much too high and the masked area becomes pretty much pure noise.

> The problem is no matter what parameter i change, it also always affect unmasked area, inverse way.

it shouldn't necessarily be in the inverse way, but yes, even though you're masking the noise and only adding extra noise in a specific area that will affect the rest of the generation. this isn't masking sampling like inpainting, _only_ the noise. the model is going to try to find a way to make whatever you did fit in with the rest of the image it's generating which will have a global effect.

one way to try to minimize global effects is to schedule the changes to start later on but this also gives the model less time to deal with stuff like weird/excessive noise. for example, try using something excessively high like `factor: 1.4` but set `start_percent` to 0.8 or 0.9. not saying the result will be good or anything but may give you an idea of how adjusting the scheduling can affect the generation.

my advice is to play around with it on low-resolution images to get a sense of how different values change things. you can start with the example workflows i linked and switch to your own model/prompts.

hope this was helpful! if you still have questions, feel free to ask.
